src/platform/backends/shared_linux/process.cpp

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (C) 2019 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "process.h"
+#include "process_spec.h"
+
+#include <multipass/logging/log.h>
+
+namespace mp = multipass;
+namespace mpl = multipass::logging;
+
+mp::Process::Process(std::unique_ptr<mp::ProcessSpec>&& spec) : process_spec{std::move(spec)}
+{
+    setProgram(process_spec->program());
+    setArguments(process_spec->arguments());
+    setProcessEnvironment(process_spec->environment());
+    if (!process_spec->working_directory().isNull())
+        setWorkingDirectory(process_spec->working_directory());
+
+    setProcessChannelMode(QProcess::ForwardedErrorChannel); // default to forwarding child stderr to console
+}
+
+void mp::Process::start(const QStringList& extra_arguments)
+{
+    QProcess::start(program(), arguments() << extra_arguments);
+}
+
+bool mp::Process::run_and_return_status(const QStringList& extra_arguments, const int timeout)
+{
+    start(extra_arguments);
+    if (!waitForFinished(timeout))
+    {
+        mpl::log(mpl::Level::info, qPrintable(process_spec->program()), qPrintable(errorString()));
+        return false;
+    }
+
+    return exitStatus() == QProcess::NormalExit && exitCode() == 0;
```

Comment:
`exitStatus` will always be `NormalExit` here, no?

Reply:
Unless it crashes, but yeah it's not a good check. This returns true if the binary isn't even found!

I just took this line from the existing code in mp::utils::run_cmd_for_status. I didn't want to accidentally change behaviour during this refactor. But since you point it out, I'll make it more solid now.

Reply:
Wouldn't `waitForFinished()` catch those above, though?

Reply:
waitForFinished() returns true if the process exits normally, or crashes. It doesn't care about return code.
exitStatus tells us if it crashed.

I've moved this code around a tad, see what you think.